### ALUMNO09 - DANIEL GONZALEZ PALAZON

### Ejercicio 3.1

### Reduction operation: the sum of the numbers in the range [0, a)

In [12]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.
Time taken by reduction operation: 0.2705988883972168 seconds
264 ms ± 1.41 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [10]:
# — Ejercicio 3.2 a) —  
import time

# Crear la lista de Python con 10**6 elementos
n = 10**6
lst = list(range(n))

# a) Medir con bucle for
start = time.time()
total_for = 0
for x in lst:
    total_for += x
elapsed_for = time.time() - start

print(f"Bucle for:   total={total_for}, tiempo={elapsed_for:.4f} s")

# b) Medir con sum()
start = time.time()
total_sum = sum(lst)
elapsed_sum = time.time() - start

print(f"sum(lst):    total={total_sum}, tiempo={elapsed_sum:.4f} s")


Bucle for:   total=499999500000, tiempo=0.4389 s
sum(lst):    total=499999500000, tiempo=0.0319 s


In [11]:
# — Ejercicio 3.2 b) —  
import time
import numpy as np

# Convertir la lista a un array de NumPy
arr = np.array(lst)

# a) Medir con bucle for sobre el array
start = time.time()
total_for_np = 0
for x in arr:
    total_for_np += x
elapsed_for_np = time.time() - start

print(f"Bucle for NumPy:   total={total_for_np}, tiempo={elapsed_for_np:.4f} s")

# b) Medir con np.sum()
start = time.time()
total_sum_np = np.sum(arr)
elapsed_sum_np = time.time() - start

print(f"np.sum(arr):       total={total_sum_np}, tiempo={elapsed_sum_np:.4f} s")


Bucle for NumPy:   total=499999500000, tiempo=0.9469 s
np.sum(arr):       total=499999500000, tiempo=0.0029 s


### 3.2 c) Análisis de resultados

Al ejecutar los tres procedimientos de suma en el rango \[0, 10<sup>6</sup>), hemos obtenido estos tiempos aproximados:

| Método                              | Tiempo        |
|-------------------------------------|---------------|
| 1. Código original (`reduc_operation`)       | 0.27 s        |
| 2.a. Bucle `for` sobre lista        | 0.44 s        |
| 2.b. `sum(lst)`                     | 0.03 s        |
| 3.a. Bucle `for` sobre `ndarray`    | 0.95 s        |
| 3.b. `np.sum(arr)`                  | 0.003 s       |

**Puntos clave:**

1. **Código original vs. bucle sobre lista**  
   - Ambos usan un bucle en Python puro para acumular elemento a elemento.  
   - La función genérica `reduc_operation` sobre `range(a)` tarda ~0.28 s, mientras que la suma manual de `list(range(n))` tarda ~0.44 s.  
   - La diferencia (~0.16 s) se debe al coste de creación de la lista completa antes de iterar y a la llamada a la función.

2. **`sum(lst)`**  
   - Implementada en C dentro del intérprete de Python, evita el bucle en Python y recorre la lista con código nativo.  
   - Mejora el rendimiento en un orden de magnitud: de ~0.44 s a ~0.03 s.

3. **Bucle `for` sobre `ndarray`**  
   - Aunque el tipo de datos cambia, se sigue iterando en Python elemento a elemento, lo que introduce *más* overhead que con listas, porque cada acceso a un elemento de `ndarray` implica comprobaciones adicionales de NumPy.  
   - Por eso la suma manual sobre `arr = np.array(lst)` es aún más lenta (~0.95 s).

4. **`np.sum(arr)`**  
   - Ejecuta la operación de reducción íntegramente en C (y optimizado con vectores SIMD si el hardware lo permite), accediendo a memoria contigua.  
   - Resultado: la ejecución es ultrarrápida (~0.003 s), casi 100× más rápida que `sum(lst)` y ~300× más rápida que el bucle Python sobre `ndarray`.

---

**Conclusión:**  
- El principal cuello de botella en Python puro es el bucle interpretado elemento a elemento.  
- Aprovechar implementaciones nativas (ya sea `sum()` para listas o `np.sum()` para arrays) traslada la carga al código compilado en C, consiguiendo enormes mejoras de rendimiento.
